src/components/cylc/tree/Tree.vue

```diff
@@ -45,36 +105,102 @@ export default {
-      collapseFilter: null
+      collapseFilter: null,
+      tasksFilter: {
+        name: '',
+        states: []
+      },
+      filtered: false
     }
   },
-  created () {
-    TreeEventBus.$on('tree-item-created', this.onTreeItemCreated)
-    TreeEventBus.$on('tree-item-destroyed', this.onTreeItemDestroyed)
-    TreeEventBus.$on('tree-item-expanded', this.onTreeItemExpanded)
-    TreeEventBus.$on('tree-item-collapsed', this.onTreeItemCollapsed)
-    TreeEventBus.$on('tree-item-clicked', this.onTreeItemClicked)
+  computed: {
+    taskStates: () => {
+      return TaskState.enumValues.map(taskState => {
+        return {
+          text: taskState.name.toLowerCase().replace(/_/g, ' '),
+          value: taskState.name
+        }
+      }).sort((left, right) => {
+        return left.text.localeCompare(right.text)
+      })
+    },
+    tasksFilterStates: function () {
+      return this.tasksFilter.states.map(selectedTaskState => {
+        return selectedTaskState.value.toLowerCase()
+      })
+    }
   },
-  destroyed () {
-    // we cannot simply call TreeEventBus.$off() as we may have more trees listening...
-    TreeEventBus.$off('tree-item-created', this.onTreeItemCreated)
-    TreeEventBus.$off('tree-item-destroyed', this.onTreeItemDestroyed)
-    TreeEventBus.$off('tree-item-expanded', this.onTreeItemExpanded)
-    TreeEventBus.$off('tree-item-collapsed', this.onTreeItemCollapsed)
-    TreeEventBus.$off('tree-item-clicked', this.onTreeItemClicked)
+  watch: {
+    workflows: {
+      deep: true,
+      handler: function (val, oldVal) {
+        if (this.filtered) {
+          this.$nextTick(() => {
+            this.filterTasks()
+          })
+        }
+      }
+    }
   },
   methods: {
+    filterByTaskName () {
+      return this.tasksFilter.name && this.tasksFilter.name.trim() !== ''
+    },
+    filterByTaskState () {
+      return this.tasksFilter.states && this.tasksFilter.states.length > 0
+    },
+    filtersEnabled () {
+      return this.filterByTaskName() || this.filterByTaskState()
+    },
+    filterTasks () {
+      if (this.filtersEnabled()) {
+        this.filterNodes(this.workflows)
+        this.filtered = true
+      } else {
+        this.removeAllFilters()
+        this.filtered = false
+      }
+    },
+    filterNodes (nodes) {
+      for (const node of nodes) {
+        this.filterNode(node)
+      }
+    },
+    filterNode (node) {
+      let filtered = false
+      if (['cyclepoint', 'family-proxy'].includes(node.type)) {
+        for (const child of node.children) {
+          filtered = this.filterNode(child) || filtered
+        }
+      } else if (node.type === 'task-proxy') {
+        if (this.filterByTaskName() && this.filterByTaskState()) {
+          filtered = node.node.name.includes(this.tasksFilter.name) && this.tasksFilterStates.includes(node.node.state)
+        } else if (this.filterByTaskName()) {
+          filtered = node.node.name.includes(this.tasksFilter.name)
```

Comment:
It would be good to upgrade this plain text search to a regex to match current `cylc review` functionality as task names can become very complex.